File: Assignment 5/Rod-Cutting-Problem/top_down.cpp
#include<bits/stdc++.h>

using namespace std;

int max(int a, int b) {
    return (a > b) ? a : b;
}

int Memoized_cut_rod_Array(vector<int> price, int n, int r[]) {
    
    int max_value;

    if(r[n+1] >= 0) {
        return r[n+1];
    }

    if(n==0) {
        max_value = 0;
    }

    else
    {
        max_value == INT_MIN;

        for (int i = 1; i <= n; i++) {
            max_value = max(max_value, (price[i] + Memoized_cut_rod_Array(price, n-i, r)));
        }
    }

    r[n+1] = max_value;

    return max_value;
}

int Memoized_cut_rod(vector<int> price, int n) {
    int r[n+1];

    for (int i = 0; i <= n; i++)
    {
        r[i] = INT_MIN;
    }
    
    return Memoized_cut_rod_Array(price, n, r);
}

int main()
{
    vector<int> price{1, 5, 8, 9, 10, 17, 17, 20};

    cout << "Maximum value : " << Memoized_cut_rod(price, price.size()) << "\n";
    return 0;
}

File: Assignment 5/Rod-Cutting-Problem/top_down.py
def Cut_Rod(p, n):
    r = [0]*(n+1)
    for i in range(0, n+1):
        r[i] = -1
    return Cut_Rod_Aux(p, n, r)


def Cut_Rod_Aux(p, n, r):
    if r[n] >= 0:
        return r[n]
    if n == 0:
        max_value = 0
    else:
        max_value = -1
        for i in range(1, n+1):
            max_value = max(max_value, p[i] + Cut_Rod_Aux(p, n-i, r))
        r[n] = max_value
    return max_value


p = [0, 1, 4, 9, 10, 17, 17, 19, 20, 25]  # 0 value is of no use

print("Maximum value : " + str(Cut_Rod(p, 9)))